src/main/kotlin/com/github/avrokotlin/avro4k/encoder/ListEncoder.kt

```diff
@@ -1,78 +1,30 @@
 package com.github.avrokotlin.avro4k.encoder
 
+import com.github.avrokotlin.avro4k.AvroConfiguration
 import kotlinx.serialization.ExperimentalSerializationApi
 import kotlinx.serialization.descriptors.SerialDescriptor
-import kotlinx.serialization.encoding.AbstractEncoder
-import kotlinx.serialization.encoding.CompositeEncoder
 import kotlinx.serialization.modules.SerializersModule
 import org.apache.avro.Schema
+import org.apache.avro.generic.GenericArray
 import org.apache.avro.generic.GenericData
-import org.apache.avro.generic.GenericFixed
-import java.nio.ByteBuffer
 
 @ExperimentalSerializationApi
-class ListEncoder(private val schema: Schema,
-                  override val serializersModule: SerializersModule,
-                  private val callback: (GenericData.Array<Any?>) -> Unit) : AbstractEncoder(), StructureEncoder {
-
-   private val list = mutableListOf<Any?>()
-
-   override fun endStructure(descriptor: SerialDescriptor) {
-      val generic = GenericData.Array(schema, list.toList())
-      callback(generic)
-   }
-
-   override fun fieldSchema(): Schema = schema.elementType
-
-   override fun beginStructure(descriptor: SerialDescriptor): CompositeEncoder {
-      return super<StructureEncoder>.beginStructure(descriptor)
-   }
-
-   override fun addValue(value: Any) {
-      list.add(value)
-   }
-
-   override fun encodeString(value: String) {
-      list.add(StringToAvroValue.toValue(schema, value))
-   }
-
-   override fun encodeLong(value: Long) {
-      list.add(value)
-   }
-
-   override fun encodeDouble(value: Double) {
-      list.add(value)
-   }
-
-   override fun encodeBoolean(value: Boolean) {
-      list.add(value)
-   }
-
-   override fun encodeShort(value: Short) {
-      list.add(value)
-   }
-
-   override fun encodeByteArray(buffer: ByteBuffer) {
-      list.add(buffer)
-   }
-
-   override fun encodeFixed(fixed: GenericFixed) {
-      list.add(fixed)
-   }
-
-   override fun encodeByte(value: Byte) {
-      list.add(value)
-   }
-
-   override fun encodeFloat(value: Float) {
-      list.add(value)
-   }
-
-   override fun encodeInt(value: Int) {
-      list.add(value)
-   }
-
-   override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) {
-      list.add(ValueToEnum.toValue(fieldSchema(), enumDescriptor, index))
-   }
+class ListEncoder(
+        listSchema: Schema,
+        size: Int,
+        override val serializersModule: SerializersModule,
+        override val configuration: AvroConfiguration,
+        private val callback: (GenericArray<Any?>) -> Unit,
+) : AvroStructureEncoder() {
+    override val currentUnresolvedSchema: Schema = listSchema.elementType
+
+    private val list = GenericData.Array<Any?>(size, listSchema)
```

Comment:
Little perf improvement: Since we use directly `GenericData.Array`, apache avro library won't need to copy the list content to a GenericData.Array, we do it directly, and with an array size to not grow the array for nothing